src/backend/libpq/rangerrest.c

```diff
@@ -245,279 +257,326 @@ json_object *create_ranger_request_json_batch(List *args)
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+/**
+ * @returns: 0 curl success; -1 curl failed
+ */
+int call_ranger_rest(CURL_HANDLE curl_handle, const char* request)
 {
-    CURLcode res;
-    Assert(request != NULL);
-
-    curl_global_init(CURL_GLOBAL_ALL);
-
-    /* init the curl session */
-    curl_handle->curl_handle = curl_easy_init();
-    if (curl_handle->curl_handle == NULL)
-    {
-        goto _exit;
-    }
-
-    /* timeout */
-    // curl_easy_setopt(curl_handle, CURLOPT_TIMEOUT, 1);
-
-    /* specify URL to get */
-    //curl_easy_setopt(curl_handle->curl_handle, CURLOPT_URL, "http://localhost:8089/checkprivilege");
-    StringInfoData tname;
-    initStringInfo(&tname);
-    appendStringInfo(&tname, "http://");
-    appendStringInfo(&tname, rps_addr_host);
-    appendStringInfo(&tname, ":");
-    appendStringInfo(&tname, "%d", rps_addr_port);
-    appendStringInfo(&tname, "/rps");
-    curl_easy_setopt(curl_handle->curl_handle, CURLOPT_URL, tname.data);
-
-    /* specify format */
-    // struct curl_slist *plist = curl_slist_append(NULL, "Content-Type:application/json;charset=UTF-8");
-    // curl_easy_setopt(curl_handle, CURLOPT_HTTPHEADER, plist);
-
-
-    //curl_easy_setopt(curl_handle->curl_handle, CURLOPT_POSTFIELDSIZE_LARGE, 1000);
-    //curl_easy_setopt(curl_handle->curl_handle, CURLOPT_HTTPGET, 0);
-    //curl_easy_setopt(curl_handle->curl_handle, CURLOPT_CUSTOMREQUEST, "POST");
-
-    struct curl_slist *headers = NULL;
-    //curl_slist_append(headers, "Accept: application/json");
-    headers = curl_slist_append(headers, "Content-Type:application/json");
-    curl_easy_setopt(curl_handle->curl_handle, CURLOPT_HTTPHEADER, headers);
-
-    //curl_easy_setopt(curl_handle->curl_handle, CURLOPT_POST, 1L);
-    curl_easy_setopt(curl_handle->curl_handle, CURLOPT_POSTFIELDS,request);
-    //"{\"requestId\": 1,\"user\": \"hubert\",\"clientIp\":\"123.0.0.21\",\"context\": \"SELECT * FROM sales\",\"access\":[{\"resource\":{\"database\":\"a-database\",\"schema\":\"a-schema\",\"table\":\"sales\"},\"privileges\": [\"select\"]}]}");
-    /* send all data to this function  */
-    curl_easy_setopt(curl_handle->curl_handle, CURLOPT_WRITEFUNCTION, write_callback);
-    curl_easy_setopt(curl_handle->curl_handle, CURLOPT_WRITEDATA, (void *)curl_handle);
-
-    res = curl_easy_perform(curl_handle->curl_handle);
-
-    /* check for errors */
-    if(res != CURLE_OK)
-    {
-        elog(WARNING, "curl_easy_perform() failed: %s\n",
-                curl_easy_strerror(res));
-    }
-    else
-    {
-        elog(LOG, "%d bytes retrieved from Ranger Restful API.",
-                curl_handle->response.size);
-    }
+	int ret = -1;
+	CURLcode res;
+	Assert(request != NULL);
+
+	curl_global_init(CURL_GLOBAL_ALL);
+
+	/* init the curl session */
+	curl_handle->curl_handle = curl_easy_init();
+	if (curl_handle->curl_handle == NULL)
+	{
+		goto _exit;
+	}
+
+	// hard-coded timeout
+	curl_easy_setopt(curl_handle->curl_handle, CURLOPT_TIMEOUT, 30L);
```

Comment:
Currently we could define a macro to replace this magic number.  
We could investigate whether this timeout should be a guc laterly.